### PR TITLE
DMIC: eliminate the gain ramping task

### PR DESCRIPTION
**Pull request by lyakh**
The DMIC driver currently uses a separate task to perform gain ramping on start up. This can be achieved from the `.copy()` context avoiding an additional asynchronous context and races associated with it.

### CODE REVIEW

**Comment by marcinszkudlinski** [comment]
@juimonen, pls take a look

File: src/drivers/intel/dmic/dmic.c [hunk 1]
```diff
@@ -59,19 +52,29 @@ static const uint32_t base[4] = {PDM0, PDM1, PDM2, PDM3};
...
+	 */
+
+	dai_dbg(dai, "dmic_gain_ramp()");
 
 	key = k_spin_lock(&dai->lock);
 
+	if (dmic->gain == DMIC_HW_FIR_GAIN_MAX << 11) {
+		k_spin_unlock(&dai->lock, key);
+		return;
+	}
```
Comment (singalsu) [inline src/drivers/intel/dmic/dmic.c:76]: Could this check and unlock be before spinlock, then just return is enough?

File: src/drivers/intel/dmic/dmic.c [hunk 2]
```diff
@@ -369,16 +372,6 @@ static void dmic_start(struct dai *dai)
 	dmic->state = COMP_STATE_ACTIVE;
 	k_spin_unlock(&dai->lock, key);
 
-	/* Currently there's no DMIC HW internal mutings and wait times
-	 * applied into this start sequence. It can be implemented here if
-	 * start of audio capture would contain clicks and/or noise and it
-	 * is not suppressed by gain ramp somewhere in the capture pipe.
-	 */
-
-	schedule_task(&dmic->dmicwork, DMIC_UNMUTE_RAMP_US,
```
Comment (singalsu) [inline src/drivers/intel/dmic/dmic.c:378]: Variable dmic->gain_coef is computed from DMIC_UNMUTE_RAMP_US. Would you also update the equation in dmic.h and dmic.c for actual rate of this callback? This works only for 1 ms now.

File: src/drivers/intel/dmic/dmic.c (hunk 1 above)
Comment (lyakh) [reply src/drivers/intel/dmic/dmic.c:76]: The same DAI cannot run on multiple cores, so, I think - yes, we can rely on the fact, that as long as the pipeline task is running, it is safe to access the `dmic` object for reading

File: src/drivers/intel/dmic/dmic.c (hunk 2 above)
Comment (singalsu) [reply src/drivers/intel/dmic/dmic.c:378]: Does it work for event driven scheduling?

**Comment by wszypelt** [comment]
On internal testing. Please be patient, there are problems with the internal service, we will try to fix it within 24 hours

**Review by lgirdwood — COMMENTED** [review]
Can we simplify and do the ramp at the start of copy() instead of adding another callback ?

**Comment by marcinszkudlinski** [comment]
> Can we simplify and do the ramp at the start of copy() instead of adding another callback ?
@lgirdwood, Dmic driver code is not involved in copy operation, so there's nothing like "start of copy()" in dmic.c code.

**Comment by lgirdwood** [comment]
> > Can we simplify and do the ramp at the start of copy() instead of adding another callback ?
> > @lgirdwood, Dmic driver code is not involved in copy operation, so there's nothing like "start of copy()" in dmic.c code.

I think it makes sense here to add copy() to the DAIs (like everything else) and keep the interface simple (and saves an extra call)

**Comment by marcinszkudlinski** [comment]
> > > Can we simplify and do the ramp at the start of copy() instead of adding another callback ?
> > > @lgirdwood, Dmic driver code is not involved in copy operation, so there's nothing like "start of copy()" in dmic.c code.
> 
> I think it makes sense here to add copy() to the DAIs (like everything else) and keep the interface simple (and saves an extra call)

It is just a matter of name, .copy or .pre_copy - would be called anyway at the same moment. The point is - this method does not do any copying, it is a dma/dai job
.pre_copy maybe a more accurate name in fact

**Comment by lgirdwood** [comment]
> > > > Can we simplify and do the ramp at the start of copy() instead of adding another callback ?
> > > > @lgirdwood, Dmic driver code is not involved in copy operation, so there's nothing like "start of copy()" in dmic.c code.
> > 
> > 
> > I think it makes sense here to add copy() to the DAIs (like everything else) and keep the interface simple (and saves an extra call)
> 
> It is just a matter of name, .copy or .pre_copy - would be called anyway at the same moment. The point is - this method does not do any copying, it is a dma/dai job .pre_copy maybe a more accurate name in fact

Agree, copy() naming does not align with usage today - but soon this call will be updated to the module C API and be called process() and this is a better naming/meaning match. Adding a new API adds a bit more complexity when it comes to the module API and I'd like to avoid and reuse (where good mapping permits).

**Comment by lyakh** [comment]
> > > > > Can we simplify and do the ramp at the start of copy() instead of adding another callback ?
> > > > > @lgirdwood, Dmic driver code is not involved in copy operation, so there's nothing like "start of copy()" in dmic.c code.
> > > 
> > > 
> > > I think it makes sense here to add copy() to the DAIs (like everything else) and keep the interface simple (and saves an extra call)
> > 
> > 
> > It is just a matter of name, .copy or .pre_copy - would be called anyway at the same moment. The point is - this method does not do any copying, it is a dma/dai job .pre_copy maybe a more accurate name in fact
> 
> Agree, copy() naming does not align with usage today - but soon this call will be updated to the module C API and be called process() and this is a better naming/meaning match. Adding a new API adds a bit more complexity when it comes to the module API and I'd like to avoid and reuse (where good mapping permits).

@lgirdwood but I should call it `.copy()` now, not `.process()` directly?

**Comment by lgirdwood** [comment]
> @lgirdwood but I should call it `.copy()` now, not `.process()` directly?

Yes, call it copy() today - there will be a subsequent update that then renames later and aligns with the module API.

File: src/drivers/intel/dmic/dmic.c (hunk 2 above)
Comment (lyakh) [reply src/drivers/intel/dmic/dmic.c:378]: @singalsu Scheduling can be timer- or DMA-based with SOF. DMA-based scheduling is used by non-Intel platforms, by older Intel platforms like BYT, and in some rare cases, like with codec clock sourcing. So, in CI, I think DMA scheduling is only used with pre-cAVS Intel platforms with with non-Intel platforms. Do any of them use DMIC?

Concerning non-default scheduling period, when the timer-driven pipeline has a non-1ms period. This is less clear to me. You're saying that if, say, a pipeline would be scheduled every 10ms, that would be too coarse for DMIC ramping? But is this ever done with DMIC pipelines? And if for some reason someone does decide to have such a scheduling period, maybe a more coarse ramping is acceptable to them?

**Comment by lgirdwood** [comment]
@lyakh @singalsu do we expect the gain ramp to change (and fail tests) ? we are seeing a failure on the DMIC tests on internal CI.

**Comment by singalsu** [comment]
> @lyakh @singalsu do we expect the gain ramp to change (and fail tests) ? we are seeing a failure on the DMIC tests on internal CI.

The capture ramps have been the same for a long time. If something has changed there could be a new bug. But I don't think CI tests the ramp shape, it can only detect no audio or much too silent audio kind of cases.

File: src/drivers/intel/dmic/dmic.c (hunk 2 above)
Comment (singalsu) [reply src/drivers/intel/dmic/dmic.c:378]: Not sure I follow, but you can test that you see a ramp that has the length that is described in topology in DAI DMIC parameters (IPC3) or get the correct hard coded ramp time (IPC4). Change pipeline scheduling to e.g. 2 ms and check that you still get correct ramp time. We could address the zipper noise with long period pipelines later. It's a subjective issue. If the pipeline is for machine hearing it doesn't matter.

**Comment by lgirdwood** [comment]
Rerun CI as not expecting a build failure on only 1 platform.

**Comment by lgirdwood** [comment]
SOFCI TEST

File: src/include/sof/lib/dai.h [hunk 3]
```diff
@@ -86,6 +86,7 @@ struct dai_ops {
 	int (*remove)(struct dai *dai);
 	uint32_t (*get_init_delay_ms)(struct dai *dai);
 	int (*get_fifo_depth)(struct dai *dai, int direction);
+	void (*copy)(struct dai *dai);
```
Comment (kv2019i) [inline src/include/sof/lib/dai.h:89]: A bit of documentation would be nice. In the context of the dmic problem this is easily understandable, but if you zoom out a bit, a copy method with no arguments looks very odd. Copy what and where?

**Review by kv2019i — APPROVED** [review]
@lyakh Documentation of the new copy() method could be improved in the first patch, but otherwise this looks great.

**Comment by lgirdwood** [comment]
@mmaka1 did you have any comments here - is this OK as part of the path forward to a copier API ?

**Comment by lyakh** [comment]
@lgirdwood CI on-device are 2 unrelated suspend -resume failures: https://sof-ci.01.org/sofpr/PR5852/build602/devicetest/?model=TGLU_UP_HDA_ZEPHYR&testcase=check-suspend-resume-5 and https://sof-ci.01.org/sofpr/PR5852/build602/devicetest/?model=WHL_UPEXT_HDA_ZEPHYR&testcase=check-suspend-resume-with-capture-5, checkpatch is complaining about a line, ending in '(' which I agree isn't pretty but looked like the best option in that case to avoid way too long a line

**Comment by lgirdwood** [comment]
Merging since this fixes an issue, we can align to copier APIs incrementally.

**Pull request merged**